# Milestones 1

Nama : Reyki Seprianza

Batch : ftds 09

[Link heroku](https://reyki-seprianza-ftds-009-p0m1.herokuapp.com)

reyki-seprianza-ftds-009-p0m1.herokuapp.com

## ***file ipynb ini adalah master sebelum di pecah untuk di deploy ke heroku***

*Milestones ini dibuat guna mengevaluasi pembelajaran pada Hacktiv8 Data Science Fulltime Program khususnya pada Phase 0.*


--------------------------------------------------------------------------

## Dataset Description
Unduh dataset [disini](https://www.kaggle.com/aungpyaeap/supermarket-sales).


## Assignment Instructions

*Milestones 1* dikerjakan dalam bentuk dashboard Streamlit dengan beberapa **kriteria wajib** di bawah ini:

1. *Library* utama yang digunakan adalah **Streamlit**.

2. *Project* dinyatakan selesai dan diterima untuk dinilai jika sudah di-deploy ke Heroku.

3. Dashboard yang dibuat terdiri dari minimal 2 layout : `Visualisasi` dan `Hypothesis Testing`.

4. Untuk layout `Visualisasi` :
   1. Minimal ada 4 figure yang ditampilkan dalam halaman.
   2. Ada callback untuk interactivity figure yang telah dibuat.

5. Untuk layout `Hypothesis Testing` :
   1. Buatlah 1 hypothesis testing terhadap dataset yang ada.
   2. Tulis proses testing serta hasil analisa dalam bentuk Markdown pada layout hypothesis testing.

6. Tetap lakukan cleaning dan preprocessing pada data. Data yang ditampilkan pada dashboard adalah data clean yang telah melalui proses preprocessing.

7. Deploy dashboard yang telah dibuat ke heroku dengan format url berupa `nama_student-batch-p0m1.herokuapp.com`, misal `reyki-seprianza-ftds-009-p0m1.herokuapp.com`.

8. Presentasikan dashboard yang telah dibuat pada P1W1D4PM.


## Assignment Submission

1. Simpan assignment pada sesi ini dengan nama `h8dsft_Milestone1_<nama-student>.ipynb`, misal `h8dsft_Milestone1_reyki_seprianza.ipynb`.

2. Pada file `.ipynb` poin 1 diatas, tambahkan URL deployment Heroku masing-masing student. Tidak adanya URL Heroku di file `.ipynb` akan menyebabkan tidak dinilainya deployment Streamlit.

3. File yang berkaitan dengan Streamlit juga wajib diupload. Akan lebih bagus jika file deployment Streamlit masuk ke dalam folder yang berbeda dengan file `.ipynb`. Contoh hasil commit tugas yang baik : 
   * `h8dsft_Milestone1_reyki_seprianza.ipynb`
   * `/deployment`
     - `requirements.txt`
     - `Procfile`
     - `app.py`
     - etc

4. Ketidaklengkapan jawaban yang diupload, seperti hanya melampirkan file `ipynb` saja atau hanya melampirkan file deployment Streamlit saja, akan menyebabkan pengurangan nilai yang signifikan pada tugas kali ini.

5. Push Assigment yang telah dibuat ke akun Github masing-masing student.

## Assignment Objectives

*Milestones 1* ini dibuat guna mengevaluasi Pembelajaran Phase 0 sebagai berikut:

- Mampu melakukan hypothesis testing terhadap dataset yang diberikan.
- Mampu membuat dashboard interaktif terhadap dataset yang diberikan.
- Mampu mempresentasikan hasil visualisasi data dan temuan.

---

## Assignment Rubrics

### Code Review

| Criteria|Meet Expectations|Points|
| --- | --- | --- |
| Streamlit Figure | Mampu membuat minimal 4 figure dengan Streamlit | 20 pts each (Max 80) |
| Streamlit Callback | Mampu membuat callback dengan Streamlit | 50 pts |
| Hypothesis Testing | Mampu melakukan hypothesis testing dan menampilkannya dengan Streamlit | 50 pts |

### Readability

| Criteria | Meet Expectations | Points|
| --- | --- | --- |
| Tertata Dengan Baik | Semua baris kode terdokumentasi dengan baik dengan menggunakan Markdown untuk penjelasan kode. | 10 pts |

### Analysis

| Criteria | Meet Expectations | Points |
| --- | --- | --- |
| Overall Analysis | Menarik informasi/kesimpulan dari keseluruhan eksplorasi yang dilakukan. | 10 pts |

---

```
Total Points : 200

Catatan : Penilaian Milestone  juga dapat dipengaruhi oleh aktivitas student selama Phase 0 berlangsung, baik sesi kelas maupun sesi mentoring dengan buddy-nya masing-masing sehingga terdapat kemungkinan adanya penambahan atau pengurangan nilai diluar rubric yang telah disebutkan diatas.
```

------------------------------------------------------------------------------------------

# Background Story

Perkenalkan saya adalah **Reyki Seprianza** seorang data scientist dari perusahaan retail yang cukup ternama di Myanmar.

Saya tergabung dalam tim Market Research & Development diberikan sebuah datasheet yang berisikan data penjualan toko perusahaan kami selama 3 bulan (Quartal 1) yang berada di kota Naypytaw, Yangon dan Mandalay.

Saya diminta untuk memberikan hasil visualisasi data yang dimiliki teritama mengenai analisa konsumen dan memberikan penjabaran hipotesa untuk mendukung tim Market Research & Development dalam membantu memberikan keputusan kepada perusahaan untuk membuat sebuah food court disalah satu kota yang ditargetkan. Pertimbangan dalam masa pandemi maka tim Market Research & Development mencoba mengalihkan pemakaian uang tunai dan kartu yang paling beresiko terpapar droplet yaitu dengan menggunakan e-wallet. Sehingga food court yang ditargetkan adalah yang berada di kota-kota dengan pengguna e-wallet terbanyak. Berdasarkan hasil analisis data yang dilakukan maka kota yang memiliki populasi penggunaan e-wallet terbanyak adalah kota Yangon.



Maka dari itu projek ini dibuat agar tim Market Research & Development dapat mengambil keputusan bagi perkembangan perusahaan kedepannya.

Secara sekilas kota Yangon adalah  kota dengan penggunaan Ewallet terbanyak, 
Objectif kita adalah untuk membuktikan apakah rata-rata pengguna Ewallet di yangon berbeda signifikan dengan kota lainnya.


🔍**Objektif kita** 


    >  untuk membuktikan apakah rata-rata pengguna Ewallet di Yangon berbeda signifikan dengan kota lainnya.

    
Hipotesa yang akan di buktikan pada kesempatan ini adalah

>h0  μ Pengguna Ewallet di kota Yangon = μ Pengguna Ewallet di kota Naypyitaw dan Mandalay


>h1  μ Pengguna Ewallet di kota Yangon != μ Pengguna Ewallet di kota Naypyitaw dan Mandalay



## Work Space

In [1]:
#import library yg di butuhkan
import pandas as pd
import matplotlib.pyplot as plt
from scipy import stats
import seaborn as sns
import numpy as np
import plotly.express as px
import plotly.graph_objects as go

import warnings
warnings.filterwarnings('ignore') #abaikan warning

In [2]:
#Data loading
df = pd.read_csv('deployment/supermarket_sales.csv')
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Invoice ID               1000 non-null   object 
 1   Branch                   1000 non-null   object 
 2   City                     1000 non-null   object 
 3   Customer type            1000 non-null   object 
 4   Gender                   1000 non-null   object 
 5   Product line             1000 non-null   object 
 6   Unit price               1000 non-null   float64
 7   Quantity                 1000 non-null   int64  
 8   Tax 5%                   1000 non-null   float64
 9   Total                    1000 non-null   float64
 10  Date                     1000 non-null   object 
 11  Time                     1000 non-null   object 
 12  Payment                  1000 non-null   object 
 13  cogs                     1000 non-null   float64
 14  gross margin percentage  

> ### Attribute information

>Invoice id: Computer generated sales slip invoice identification number

>Branch: Branch of supercenter (3 branches are available identified by A, B and C).

>City: Location of supercenters

>Customer type: Type of customers, recorded by Members for customers using member card and Normal for without member card.

>Gender: Gender type of customer

>Product line: General item categorization groups - Electronic accessories, Fashion accessories, Food and beverages, Health and beauty, Home and lifestyle, Sports and travel

>Unit price: Price of each product in $

>Quantity: Number of products purchased by customer

>Tax: 5% tax fee for customer buying

>Total: Total price including tax

>Date: Date of purchase (Record available from January 2019 to March 2019)

>Time: Purchase time (10am to 9pm)

>Payment: Payment used by customer for purchase (3 methods are available – Cash, Credit card and Ewallet)

>COGS: Cost of goods sold

>Gross margin percentage: Gross margin percentage

>Gross income: Gross income

>Rating: Customer stratification rating on their overall shopping experience (On a scale of 1 to 10)

In [4]:
df.describe().T

count        mean           std        min  \
Unit price               1000.0   55.672130  2.649463e+01  10.080000   
Quantity                 1000.0    5.510000  2.923431e+00   1.000000   
Tax 5%                   1000.0   15.379369  1.170883e+01   0.508500   
Total                    1000.0  322.966749  2.458853e+02  10.678500   
cogs                     1000.0  307.587380  2.341765e+02  10.170000   
gross margin percentage  1000.0    4.761905  6.131498e-14   4.761905   
gross income             1000.0   15.379369  1.170883e+01   0.508500   
Rating                   1000.0    6.972700  1.718580e+00   4.000000   

                                25%         50%         75%          max  
Unit price                32.875000   55.230000   77.935000    99.960000  
Quantity                   3.000000    5.000000    8.000000    10.000000  
Tax 5%                     5.924875   12.088000   22.445250    49.650000  
Total                    124.422375  253.848000  471.350250  1042.650000  
cogs                     118.497500  241.760000  448.905000   993.000000  
gross margin percentage    4.761905    4.761905    4.761905     4.761905  
gross income               5.924875   12.088000   22.445250    49.650000  
Rating                     5.500000    7.000000    8.500000    10.000000

In [5]:
#Visualisasi data

fig = px.histogram(df,x= 'Customer type',color='Gender',barmode='group',
                   title='Total Jumlah Customer Member/Non member')
fig

In [6]:
fig = px.histogram(df,x= 'City',color='Gender',barmode='group',
                   title='Total Jenis Kelamin Pelanggan Berdasarkan Cabang')
fig

In [7]:
fig = px.histogram(df,x= 'Payment',barmode='group',color='Gender',
                   title='Tipe Payment Berdasarkan Jenis Kelamin ')
fig

In [8]:
fig = px.bar(df,x='Product line',y='Unit price',color='Gender',barmode='group',title='Tipe Barang yang dibeli berdasarkan Jenis Kelamin')
fig.update_layout(uniformtext_minsize=8 , xaxis_tickangle=-45)

In [9]:
fig = px.histogram(df,x= 'Payment',barmode='group',
                   title='Total tipe pembayaran ')
fig

In [10]:
px.bar(df,x='City',y='Total',color='City',animation_frame='Date',title='Perubahan Total Harga per kota berdasarkan tanggal',
      animation_group="Branch", range_y=[0,1000])

# Data Cleaning & Analysist

In [11]:
#Missing values cheking
df.info

<bound method DataFrame.info of       Invoice ID Branch       City Customer type  Gender  \
0    750-67-8428      A     Yangon        Member  Female   
1    226-31-3081      C  Naypyitaw        Normal  Female   
2    631-41-3108      A     Yangon        Normal    Male   
3    123-19-1176      A     Yangon        Member    Male   
4    373-73-7910      A     Yangon        Normal    Male   
..           ...    ...        ...           ...     ...   
995  233-67-5758      C  Naypyitaw        Normal    Male   
996  303-96-2227      B   Mandalay        Normal  Female   
997  727-02-1313      A     Yangon        Member    Male   
998  347-56-2442      A     Yangon        Normal    Male   
999  849-09-3807      A     Yangon        Member  Female   

               Product line  Unit price  Quantity   Tax 5%      Total  \
0         Health and beauty       74.69         7  26.1415   548.9715   
1    Electronic accessories       15.28         5   3.8200    80.2200   
2        Home and lifestyle 

In [12]:
for col in df.columns:
    missing_value = df[col].isnull().value_counts()
    print(missing_value)

False    1000
Name: Invoice ID, dtype: int64
False    1000
Name: Branch, dtype: int64
False    1000
Name: City, dtype: int64
False    1000
Name: Customer type, dtype: int64
False    1000
Name: Gender, dtype: int64
False    1000
Name: Product line, dtype: int64
False    1000
Name: Unit price, dtype: int64
False    1000
Name: Quantity, dtype: int64
False    1000
Name: Tax 5%, dtype: int64
False    1000
Name: Total, dtype: int64
False    1000
Name: Date, dtype: int64
False    1000
Name: Time, dtype: int64
False    1000
Name: Payment, dtype: int64
False    1000
Name: cogs, dtype: int64
False    1000
Name: gross margin percentage, dtype: int64
False    1000
Name: gross income, dtype: int64
False    1000
Name: Rating, dtype: int64



Data yang kita miliki tidak ada missing value

In [13]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

In [14]:
#Hapus Kolom yang tidak di gunakan
df_new= df.drop(['Invoice ID',	'Branch',	'Customer type'	,'Gender',	'Product line',	'Unit price',	'Quantity',	'Tax 5%',	'Date',	'Time',	'cogs',	'gross margin percentage',	'gross income',	'Rating'], axis=1)
df_new.sample(10)

City     Total  Payment
451   Mandalay  138.6630     Cash
625   Mandalay  165.6480     Cash
645     Yangon  182.9520  Ewallet
392     Yangon   80.6610  Ewallet
558     Yangon  262.4580  Ewallet
404  Naypyitaw  718.7565  Ewallet
535  Naypyitaw  122.5245  Ewallet
108  Naypyitaw  321.1110     Cash
599     Yangon  147.6720  Ewallet
407   Mandalay  288.5820     Cash

In [15]:
#Menentukan kota yang akan di bandingkan
fig = px.histogram(df,x= 'City',color='Payment',barmode='group',
                   title='Total Jenis Kelamin Pelanggan Berdasarkan Cabang')
fig

Untuk penggunaan e-wallet terbanyak secara sekilah ada di kota Yangon, tapi harus di buktikan dahulu sesuai dengan hipotesis kita


>h0  μ Pengguna Ewallet di kota Yangon = μ Pengguna Ewallet di kota Naypyitaw dan Mandalay


>h1  μ Pengguna Ewallet di kota Yangon != μ Pengguna Ewallet di kota Naypyitaw dan Mandalay

In [16]:
#Slicing prepare data setiap kota yang menggunakan pembayaran Ewallet
df_y = df_new[df_new.City=='Yangon']
df_n = df_new[df_new.City=='Naypyitaw']
df_m = df_new[df_new.City=='Mandalay']

total_y_e = df_y[df_y.Payment == 'Ewallet']
total_n_e = df_n[df_n.Payment == 'Ewallet']
total_m_e = df_m[df_m.Payment == 'Ewallet']

print('Rata - rata penggunaan Ewalet di kota Yangon adalah: ',total_y_e.mean())
print('Rata - rata penggunaan Ewalet di kota Naypyitaw adalah: ',total_n_e.mean())
print('Rata - rata penggunaan Ewalet di kota Mandalay adalah: ',total_m_e.mean())

Rata - rata penggunaan Ewalet di kota Yangon adalah:  Total    312.098167
dtype: float64
Rata - rata penggunaan Ewalet di kota Naypyitaw adalah:  Total    350.522491
dtype: float64
Rata - rata penggunaan Ewalet di kota Mandalay adalah:  Total    296.578354
dtype: float64


In [17]:
# Slicing data
total_y_e.drop(['City','Payment'], axis=1, inplace=True)
total_n_e.drop(['City','Payment'], axis=1, inplace=True)
total_m_e.drop(['City','Payment'], axis=1, inplace=True)


### Analisa menggunakan ANOVA

In [18]:
#Mencari P-Value 
f_stat,p_value = stats.f_oneway(total_y_e, total_n_e, total_m_e)
print('P-value:',p_value)


P-value: [0.23092647]


P- Value lebih dari 0.005 

In [19]:
#Slicing untuk visualisasi hasil analisa
df_v= df
gadipake = ['Cash', 'Credit card']
df_v = df_v[df_v.Payment.isin(gadipake)== False]
df_v

Invoice ID Branch       City Customer type  Gender  \
0    750-67-8428      A     Yangon        Member  Female   
3    123-19-1176      A     Yangon        Member    Male   
4    373-73-7910      A     Yangon        Normal    Male   
5    699-14-3026      C  Naypyitaw        Normal    Male   
6    355-53-5943      A     Yangon        Member  Female   
..           ...    ...        ...           ...     ...   
992  745-74-0715      A     Yangon        Normal    Male   
993  690-01-6631      B   Mandalay        Normal    Male   
994  652-49-6720      C  Naypyitaw        Member  Female   
995  233-67-5758      C  Naypyitaw        Normal    Male   
996  303-96-2227      B   Mandalay        Normal  Female   

               Product line  Unit price  Quantity   Tax 5%      Total  \
0         Health and beauty       74.69         7  26.1415   548.9715   
3         Health and beauty       58.22         8  23.2880   489.0480   
4         Sports and travel       86.31         7  30.2085   634.3785   
5    Electronic accessories       85.39         7  29.8865   627.6165   
6    Electronic accessories       68.84         6  20.6520   433.6920   
..                      ...         ...       ...      ...        ...   
992  Electronic accessories       58.03         2   5.8030   121.8630   
993     Fashion accessories       17.49        10   8.7450   183.6450   
994  Electronic accessories       60.95         1   3.0475    63.9975   
995       Health and beauty       40.35         1   2.0175    42.3675   
996      Home and lifestyle       97.38        10  48.6900  1022.4900   

          Date   Time  Payment    cogs  gross margin percentage  gross income  \
0     1/5/2019  13:08  Ewallet  522.83                 4.761905       26.1415   
3    1/27/2019  20:33  Ewallet  465.76                 4.761905       23.2880   
4     2/8/2019  10:37  Ewallet  604.17                 4.761905       30.2085   
5    3/25/2019  18:30  Ewallet  597.73                 4.761905       29.8865   
6    2/25/2019  14:36  Ewallet  413.04                 4.761905       20.6520   
..         ...    ...      ...     ...                      ...           ...   
992  3/10/2019  20:46  Ewallet  116.06                 4.761905        5.8030   
993  2/22/2019  18:35  Ewallet  174.90                 4.761905        8.7450   
994  2/18/2019  11:40  Ewallet   60.95                 4.761905        3.0475   
995  1/29/2019  13:46  Ewallet   40.35                 4.761905        2.0175   
996   3/2/2019  17:16  Ewallet  973.80                 4.761905       48.6900   

     Rating  
0       9.1  
3       8.4  
4       5.3  
5       4.1  
6       5.8  
..      ...  
992     8.8  
993     6.6  
994     5.9  
995     6.2  
996     4.4  

[345 rows x 17 columns]

In [20]:
fig_violin = go.Figure()

Cities = ['Yangon', 'Naypyitaw', 'Mandalay']

for City in Cities:
    fig_violin.add_trace(go.Violin(x=df_v['City'][df_v['City'] == City],
                            y=df_v['Total'][df_v['City'] == City],
                            name=City,
                            box_visible=True,
                            meanline_visible=True,
                            ))

fig_violin.show()

In [24]:
y_pop = np.random.normal(total_y_e.Payment.mean(),total_y_e.Payment.std(),80)
n_pop = np.random.normal(total_n_e.Payment.mean(),total_n_e.Payment.std(),80)
m_pop = np.random.normal(total_m_e.Payment.mean(),total_m_e.Payment.std(),80)
#write(f'{y_pop.std()}, {n_pop.std()}, {m_pop.std()}')

ci = stats.norm.interval(0.95, total_y_e.Payment.mean(), total_y_e.Payment.std())
fig = plt.figure(figsize=(16,5))
sns.distplot(y_pop, label='E wallet Yangon',color='orange')
sns.distplot(n_pop, label='E wallet Naypyitaw',color='blue')
sns.distplot(m_pop, label='E wallet Mandalay C',color='red')

plt.axvline(total_n_e.Payment.mean(), color='blue', linewidth=2, label='Naypyitaw mean')
plt.axvline(total_m_e.Payment.mean(), color='red',  linewidth=2, label='Mandalay mean')

plt.axvline(ci[1], color='green', linestyle='dashed', linewidth=2, label='confidence threshold of 95%')
plt.axvline(ci[0], color='green', linestyle='dashed', linewidth=2)


plt.axvline (y_pop.mean()+f_stat[0]y_pop.std(), color='black', linestyle='dashed', linewidth=2, label = 'Alternative Hypothesis')
plt.axvline (y_pop.mean()-f_stat[0]y_pop.std(), color='black', linestyle='dashed', linewidth=2)
plt.legend()

plt.show(fig)

SyntaxError: invalid syntax (Temp/ipykernel_32/1676742918.py, line 19)

 ### 👨🏻‍💻 Hasil Analisis✍🏻 
        > Di dapatkan angka P-Value = 0.23092647
        
        > P- Value lebih dari 0.005 maka H0 bisa di buktikan 
    
## Kesmipulan 🔍 
   #### Jadi *bisa di buktikan* 👨🏻‍🔬
            > Rata-rata pengguna Ewallet di kota Yangon tidak berbeda signifikan dari rata-rata pengguna Ewallet di kota Naypyitaw dan Mandalay
    